File: lx16a/test/test_encoder_filter.py
#!/usr/bin/env python

PKG='lx16a'

from lx16a.encoder import LX16AEncoderFilter

import csv
import math
import os
import rospy
import sys
import unittest

class TestEncoderFilter(unittest.TestCase):
    ''' Test cases for LX16AEncoderFilter
    '''

    def setUp(self):
        # Constants
        self.window = 10

        # Raw data produced by servo and encoder (in lx16a/data)
        self.data_filename = os.path.join(
            os.path.dirname(__file__),
            '..', 'data', 'lx16a_raw_data_05.csv')

        # File locations for persisted ML models (in lx16a/data)
        self.classifier_filename = os.path.join(
            os.path.dirname(__file__),
            '..', 'data', 'lx16a_tree_classifier.joblib')

        self.regressor_filename = os.path.join(
            os.path.dirname(__file__),
            '..', 'data', 'lx16a_tree_regressor.joblib')

    def read_n(self, filter, n):
        ''' Utility to read the first n lines from the CSV file.

        Read the first n lines from the CSV file, update the filter,
        and return the data.
        '''
        
        data = []
        with open(self.data_filename) as csvfile:
            reader = csv.reader(csvfile, delimiter=',')
            head = [next(reader) for x in range(n)]
            for row in head:
                ros_time = rospy.Time(0, int(row[0]))
                duty = int(row[1])
                pos = int(row[2])
                count = int(row[3])
                data.append([ros_time, duty, pos, count])
                filter.update(ros_time, duty, pos)
        return data

    def test_constructor(self):
        # Load the model files
        filter = LX16AEncoderFilter(
            self.classifier_filename, self.regressor_filename, self.window)

        # Values should initialise to zero
        self.assertEquals(filter.get_revolutions(), 0, 'revolutions != 0')
        self.assertEquals(filter.get_count(), 0, 'count != 0')
        self.assertEquals(filter.get_duty(), 0, 'duty != 0')
        self.assertEquals(filter.get_angular_position(), 0, 'angular_position != 0')

        # Values should initialise to tuple (0, True)
        self.assertEquals(filter.get_servo_pos(), (0, True), 'servo_pos != (0 ,True)')

        # Values should initialise to one
        self.assertEquals(filter.get_invert(), 1, 'invert != 1')

    def test_valid_region(self):
        # Load the model files
        filter = LX16AEncoderFilter(
            self.classifier_filename, self.regressor_filename, self.window)
        data = self.read_n(filter, 400)

        self.assertEquals(filter.get_revolutions(), 1, 'revolutions != 1')
        self.assertEquals(filter.get_count(), 2134, 'count != 2134')
        self.assertEquals(filter.get_duty(), 250, 'duty != 250')
        self.assertEquals(filter.get_servo_pos(), (634, True), 'servo_pos != 634')
        self.assertEquals(filter.get_invert(), 1, 'invert != 1')

    def test_invalid_region(self):
        # Load the model files
        filter = LX16AEncoderFilter(
            self.classifier_filename, self.regressor_filename, self.window)
        data = self.read_n(filter, 100)            

        expected = (1306, False)
        self.assertEquals(filter.get_servo_pos(), expected, 'servo_pos != {}'.format(expected))

    def test_angular_position(self):
        # Load the model files
        filter = LX16AEncoderFilter(
            self.classifier_filename, self.regressor_filename, self.window)
        data = self.read_n(filter, 200)

        count = filter.get_count()
        expected = 2 * math.pi * count / 1500.0 
        self.assertEquals(filter.get_angular_position(), expected, 'angular_position != {}'.format(expected))

    def test_duty(self):
        # Load the model files
        filter = LX16AEncoderFilter(
            self.classifier_filename, self.regressor_filename, self.window)
        data = self.read_n(filter, 100)

        expected = data[-1][1]
        self.assertEquals(filter.get_duty(), expected, 'duty != {}'.format(expected))

    # Disable this test as it requires a ROS node to be available...
    def notest_reset(self):
        # Load the model files
        filter = LX16AEncoderFilter(
            self.classifier_filename, self.regressor_filename, self.window)
        data = self.read_n(filter, 300)
        filter.reset(0)

        # Values should reset to zero
        self.assertEquals(filter.get_revolutions(), 0, 'revolutions != 0')
        self.assertEquals(filter.get_count(), 0, 'count != 0')
        self.assertEquals(filter.get_duty(), 0, 'duty != 0')
        self.assertEquals(filter.get_angular_position(), 0, 'angular_position != 0')

        # Values should reset to tuple (0, True)
        self.assertEquals(filter.get_servo_pos(), (0, True), 'servo_pos != (0 ,True)')

        # Values should reset to one
        self.assertEquals(filter.get_invert(), 1, 'get_invert != 1')

    def test_invert(self):
        # Load the model files
        filter = LX16AEncoderFilter(
            self.classifier_filename, self.regressor_filename, self.window)
        data = self.read_n(filter, 200)
        filter.set_invert(True)

        expected = -1
        self.assertEquals(filter.get_invert(), expected, 'invert != {}'.format(expected))

        count = filter.get_count()
        expected = 2 * math.pi * count / 1500.0
        self.assertEquals(filter.get_angular_position(), expected, 'angular_position != {}'.format(expected))

if __name__ == '__main__':
    import rosunit
    rosunit.unitrun(PKG, 'test_encoder_filter', TestEncoderFilter)

File: lx16a/test/test_encoder_filter.cpp
#include "config.h"
#include "lx16a/lx16a_encoder_filter_python.h"
#include "lx16a/lx16a_pybind11_embed.h"
#include <gtest/gtest.h>
#include <pybind11/pybind11.h>
#include <pybind11/embed.h>
#include <ros/ros.h>

namespace py = pybind11;

// Test fixture
class EncoderFilterPythonTestSuite : public ::testing::Test
{
protected:
    ~EncoderFilterPythonTestSuite() override
    {
    }

    EncoderFilterPythonTestSuite()
    {
        // Load the test data.

    }

    void SetUp() override
    {        
        switch (lx16a::getRosPythonVersion())
        {
          case 2:
          {
            classifier_filename_ = lx16a::getDataPath()
              .append("/lx16a_tree_classifier.joblib");
            regressor_filename_ = lx16a::getDataPath()
              .append("/lx16a_tree_regressor.joblib");
            break;
          }
          case 3:
          {
            classifier_filename_ = lx16a::getDataPath()
              .append("/lx16a_tree_classifier_python3.joblib");
            regressor_filename_ = lx16a::getDataPath()
              .append("/lx16a_tree_regressor_python3.joblib");
            break;
          }
          default:
            break;
        }
    }

    void TearDown() override
    {        
    }

    const uint8_t servo_id_ = 11;
    const int16_t window_ = 10;
    const std::string test_data_filename_ = lx16a::getDataPath()
        .append("/lx16a_raw_data_05.csv");

    std::string classifier_filename_;
    std::string regressor_filename_;
};


// Declare tests
TEST_F(EncoderFilterPythonTestSuite, testConstructor)
{
    std::unique_ptr<lx16a::LX16AEncoderFilter> filter(
        new lx16a::LX16AEncoderFilterPython(
            classifier_filename_,
            regressor_filename_,
            window_));
    filter->init();
    filter->add(servo_id_);

    EXPECT_EQ(filter->getRevolutions(servo_id_), 0) << "revolutions != 0";
    EXPECT_EQ(filter->getCount(servo_id_), 0) << "count != 0";
    EXPECT_EQ(filter->getDuty(servo_id_), 0) << "duty != 0";
    EXPECT_EQ(filter->getInvert(servo_id_), 1) << "invert != 1";

    int16_t position = 0;
    bool is_valid = false;
    filter->getServoPosition(servo_id_, position, is_valid);
    EXPECT_EQ(position, 0) << "position != 0";
    EXPECT_TRUE(is_valid) << "is_valid != true";
}

TEST_F(EncoderFilterPythonTestSuite, testUpdate)
{
    std::unique_ptr<lx16a::LX16AEncoderFilter> filter(
        new lx16a::LX16AEncoderFilterPython(
            classifier_filename_,
            regressor_filename_,
            window_));
    filter->init();
    filter->add(servo_id_);

    int16_t duty = 500;
    int16_t position = 0;
    bool is_valid = false;
    ros::Time start = ros::Time();

    filter->update(servo_id_, start + ros::Duration(0.0), duty, 0);
    filter->getServoPosition(servo_id_, position, is_valid);
    EXPECT_EQ(position, 0) << "position != 0";
    EXPECT_TRUE(is_valid) << "is_valid != true";
    EXPECT_EQ(filter->getRevolutions(servo_id_), 0) << "revolutions != 0";
    EXPECT_EQ(filter->getCount(servo_id_), 0) << "count != 0";
    EXPECT_EQ(filter->getDuty(servo_id_), 500) << "duty != 500";
    EXPECT_EQ(filter->getInvert(servo_id_), 1) << "invert != 1";

    filter->update(servo_id_, start + ros::Duration(0.5), duty, 300);
    filter->getServoPosition(servo_id_, position, is_valid);
    EXPECT_EQ(position, 300) << "position != 300";
    EXPECT_TRUE(is_valid) << "is_valid != true";
    EXPECT_EQ(filter->getRevolutions(servo_id_), 0) << "revolutions != 0";
    EXPECT_EQ(filter->getCount(servo_id_), 300) << "count != 300";
    EXPECT_EQ(filter->getDuty(servo_id_), 500) << "duty != 500";
    EXPECT_EQ(filter->getInvert(servo_id_), 1) << "invert != 1";

    filter->update(servo_id_, start + ros::Duration(1.0), duty, 600);
    filter->getServoPosition(servo_id_, position, is_valid);
    EXPECT_EQ(position, 600) << "position != 600";
    EXPECT_TRUE(is_valid) << "is_valid != true";
    EXPECT_EQ(filter->getRevolutions(servo_id_), 0) << "revolutions != 0";
    EXPECT_EQ(filter->getCount(servo_id_), 600) << "count != 600";
    EXPECT_EQ(filter->getDuty(servo_id_), 500) << "duty != 500";
    EXPECT_EQ(filter->getInvert(servo_id_), 1) << "invert != 1";

    filter->update(servo_id_, start + ros::Duration(1.5), duty, 900);
    filter->getServoPosition(servo_id_, position, is_valid);
    EXPECT_EQ(position, 900) << "position != 900";
    EXPECT_TRUE(is_valid) << "is_valid != true";
    EXPECT_EQ(filter->getRevolutions(servo_id_), 0) << "revolutions != 0";
    EXPECT_EQ(filter->getCount(servo_id_), 900) << "count != 900";
    EXPECT_EQ(filter->getDuty(servo_id_), 500) << "duty != 500";
    EXPECT_EQ(filter->getInvert(servo_id_), 1) << "invert != 1";

    filter->update(servo_id_, start + ros::Duration(2.0), duty, 1100);
    filter->getServoPosition(servo_id_, position, is_valid);
    EXPECT_EQ(position, 1100) << "position != 1100";
    EXPECT_TRUE(is_valid) << "is_valid != true";
    EXPECT_EQ(filter->getRevolutions(servo_id_), 0) << "revolutions != 0";
    EXPECT_EQ(filter->getCount(servo_id_), 1100) << "count != 1100";
    EXPECT_EQ(filter->getDuty(servo_id_), 500) << "duty != 500";
    EXPECT_EQ(filter->getInvert(servo_id_), 1) << "invert != 1";

    // filter->update(servo_id_, start + ros::Duration(2.5), duty, 900);
    // filter->getServoPosition(servo_id_, position, is_valid);
    // EXPECT_EQ(position, 900) << "position != 1100";
    // EXPECT_TRUE(is_valid) << "is_valid != false";
    // EXPECT_EQ(filter->getRevolutions(servo_id_), 0) << "revolutions != 0";
    // EXPECT_EQ(filter->getCount(servo_id_), 1400) << "count != 1100";
    // EXPECT_EQ(filter->getDuty(servo_id_), 500) << "duty != 500";
    // EXPECT_EQ(filter->getInvert(servo_id_), 1) << "invert != 1";
}


TEST_F(EncoderFilterPythonTestSuite, testGetPosition)
{
    std::unique_ptr<lx16a::LX16AEncoderFilter> filter(
        new lx16a::LX16AEncoderFilterPython(
            classifier_filename_,
            regressor_filename_,
            window_));
    filter->init();
    filter->add(servo_id_);

    int16_t position = 0;
    bool is_valid = false;
    filter->getServoPosition(servo_id_, position, is_valid);
    EXPECT_EQ(position, 0) << "position != 0";
    EXPECT_TRUE(is_valid) << "is_valid != true";
}

// Run tests
int main(int argc, char **argv)
{
    // Initialise the Python interpreter
    py::scoped_interpreter guard{};
    lx16a::addCmdArgsToSys(argc, argv);

    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
